### PR TITLE
GCS_MAVLink: respect txbuf flow control for FTP messages

### PR DESCRIPTION
**Pull request by brad112358**
This gives slow radio links a fighting chance of getting FTP bulk download working efficiently, even when they use a baud rate which is much higher than their current bandwidth.  The strategy is similar to what Ardupilot already does for PARAM_VALUE messages.

This should eliminate the need to disable FTP for parameter download on slow to moderate speed radio links like mLRS.  It allows removal of a hack in mLRS which results in a decrease in parameter download time for 19 Hz mLRS from 45-60 seconds to 11-17 seconds. Other mLRS message rates show similar improvement.  This should also be good news for the ELRS rc-mavlink branch.

Also, improve get_last_txbuf() to allow message flow in the unlikely event that we stop receiving radio_status messages.

This has also been tested with TBS Crossfire and SiK radios and found to have no negative impact on parameter download.

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_FTP.cpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ bool GCS_MAVLINK::send_ftp_reply(const pending_ftp &reply)
         send_banner();
     }
     WITH_SEMAPHORE(comm_chan_lock(reply.chan));
-    if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+    if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL) || get_last_txbuf() < 33) {
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_FTP.cpp:108]: 33? Looks like params use 50?

**Comment by olliw42** [comment]
I wonder if it's a good idea to put all mftp replies under txbuf control, instead of only those of the read/readburst stream. Putting also the mftp replies for "negotiating" under control may make it harder for gcses to get their timeouts right.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -827,6 +827,15 @@ float GCS_MAVLINK::telemetry_radio_rssi()
     return last_radio_status.rssi/254.0f;
 }
 
+uint8_t GCS_MAVLINK::get_last_txbuf()
```
[the remark this answers is not in the document]
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp:830]: Do you mean the magic value 100?  That's not much magic.  Do you feel strongly about this?  I suppose we could do this, but it would complicate the logic a little where we use this function so I prefer it this way.

[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS_Common.cpp:830]: > Do you feel strongly about this?

Yes.

File: libraries/GCS_MAVLink/GCS.h [hunk 3]
```diff
@@ -805,7 +806,6 @@ class GCS_MAVLINK
     // number of extra ms to add to slow things down for the radio
     uint16_t         stream_slowdown_ms;
     // last reported radio buffer percent available
```
Comment (amilcarlucas) [inline libraries/GCS_MAVLink/GCS.h:808]: This comment should be moved to the respective line on the `last_radio_status` structure

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/GCS_MAVLink/GCS.h (hunk 3 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS.h:808]: Oops.

**Comment by brad112358** [comment]
> I wonder if it's a good idea to put all mftp replies under txbuf control, instead of only those of the read/readburst stream. Putting also the mftp replies for "negotiating" under control may make it harder for gcses to get their timeouts right.

@olliw42 Did you happen to test this? (I ask just so we can know what you saw if you did.)  Either way, I think I understand why you might wonder.   It does sometimes (less than 10% in my recent tests) happen that we wait so long for buffer space that the GCS times out and ends up re-sending the first FTP message (OpenFileRO in the case of QGC).

I have some thoughts about this.  

1)  I'd ideally like to make this work reasonably well, at least for mLRS without requiring the GCSes to modify their current timeouts, though I do think they could be improved in this area.  In the case of QGC, the timeout for all expected FTP messages is 1 second which does seem a little low for low bandwidth links, especially given the bursts of traffic Ardupilot tends to send.  (PX4 does better at spreading out normal telemetry streams)

2)  If we send the response to OpenFileRO without regard to reported buffer space, there is a real risk it or other messages might be lost.  Especially if the radio involved doesn't work as hard as mLRS does to avoid this.

3)  If we send the response to OpenFileRO when buffer space is low, even if the response is not lost because of lack of space, we have just deferred the timeout problem.  The GCS is just as likely to time out while we wait for space when we try to send the first response to the BurstReadFile message which will arrive very soon after we send the response to OpenFileRO.

Instead, I think it might try limiting the delay due to txbuf to about 500ms for all message types.  If we don't have more than .5 seconds worth of data buffered, that might help.  This would, however never work with a radio bandwidth of less than about 300 bytes/second.  Note that mLRS (which currently has at least 1500 Bytes/sec).  Of course, the current GCS timeouts would also need to change for lower bandwidth and higher latency.

File: libraries/GCS_MAVLink/GCS.h (hunk 3 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS.h]: I removed the orphan comment in my recent comit

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 1 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_FTP.cpp]: 50 was arbitrary, 33 is less arbitrary.  I have changed parameter download to 33 as well in my latest commit.

We (I) had originally used 50 for bulk parameter download a few years ago when I was trying to help TBS get parameter download working reliably on Crossfire. But experience with helping Olli optimize mLRS txbuf reporting has changed my mind on what threshold is best.

The reasons are complex, but since you asked...

Olli designed a clever strategy which chooses txbuf values to report in radio_status so as to control the bandwidth used by normal streams without requiring the buffer to actually constantly fill up to a threashold percentage before the streams slow down.  This allows typical buffer use to be smaller which minimizes message latency.  But, by design, this tends to result in equilibrium being reached where reported txbuf values can remain at 50% for fairly long periods of time.  We don't want to block burst messages (FTP and parameters) for too long so it's much better, especially when this type of algorithm is used in the radio, to keep the burst message flow control threashold below the range where we don't adjust stream_slowdown_ms.  Note that since Ardupilot also slows down normal flows when FTP and parameter download are in progress, mLRS worked OK most of the time for parameter download even with the cutoff at 50.  But the GCS was timing out sometimes with FTP because it only waits for 1 second and mLRS normally samples bandwidth use at 1 second intervals.

In my tests, implementing a limit on how long we wait for txbuf to increase solved the problem even with the threshold at 50, but we don't really want arbitrary timeouts if we can avoid them so in my latest commit, I found other ways to reduce FTP response time.

Note that the root cause of all the complexity in the above discussion is the poor design of the radio status txbuf MAVLink radio buffer and flow control mechanism.  There is no way for the radio to report the size of it's buffer so that the autopilot can do a better job understanding available bandwidth and minimizing buffer overflow and latency.  And there are no guidelines on how often the radio should send txbuf updates in radio status messages.  Sending radio status at 1 Hz like SiK radios did is entirely insufficient to deal with rapidly changing bandwidth use caused by bulk data transfer.  There are also no guidelines on how autopilots should respond to txbuf reports.  To make matters worse for radios, Ardupilot and PX4 have implemented completely different mechanisms to regulate the bandwidth of normal streams.  This poor protocol design is presumably why FTP and bulk parameter download code in Ardupilot ended up depending on hardware flow control or carefully matched baud rates or hacks like artificial limits on bandwidth demand.  Ardupilot also does a bad job of spreading out stream message transmission which results in unavoidable latency with moderately low bandwidth radios when many message types are all transmitted at about the same time.  These bursts of messages make it difficult for the radio to estimate average bandwidth use over short periods which slows down response to changes in bandwidth demand.

33 should work well for radios which use strategies similar to what mLRS does and also for radios which simply report txbuf at rates high enough to prevent their buffers from overflowing.  For radios which report txbuf at a fixed 1Hz rate, this won't help, but it shouldn't hurt and didn't hurt in my tests with SiK.

**Comment by brad112358** [comment]
I've made several improvements in the latest commit and I think I've addressed all the comments.  This is working even better in my tests.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: I replaced this code with a predicate function in my last commit so hopefully this resolves your concern.

File: libraries/GCS_MAVLink/GCS_FTP.cpp [hunk 4]
```diff
@@ -97,12 +97,8 @@ void GCS_MAVLINK::handle_file_transfer_protocol(const mavlink_message_t &msg) {
 
 bool GCS_MAVLINK::send_ftp_reply(const pending_ftp &reply)
 {
-    /*
-      provide same banner we would give with old param download
-    */
-    if (ftp.need_banner_send_mask & (1U<<reply.chan)) {
-        ftp.need_banner_send_mask &= ~(1U<<reply.chan);
-        send_banner();
+    if (!last_txbuf_is_greater(33)) { // It helps avoid GCS timeout if this is less than the threashold where we slow down normal streams (<=49)
```
Comment (amilcarlucas) [inline libraries/GCS_MAVLink/GCS_FTP.cpp:100]:
```suggestion
    if (!last_txbuf_is_greater(33)) { // It helps avoid GCS timeout if this is less than the threshold where we slow down normal streams (<=49)
```

**Comment by wvarty** [comment]
Tested with:
- ExpressLRS 2.4G hardware running the `mavlink-rc` branch, with MavFTP re-enabled
- Speedybee F405 wing FC flashed with this PR
- 250Hz packet rate, and also 333Hz-full rate

Params load substantially faster than with individual param polling (as expected) on both rates.
Mission load using both MavFTP and regular.

Flashed the FC back to mainline and confirmed that the original issues still exist: FTP param load fails (due to overloading the buffer in ELRS) and it falls back to polled param mode.

Whats left to get this merged?

**Comment by timtuxworth** [comment]
Looking at this slightly differently - I wonder why the GCS needs to get _all_ the parameters up front in order for the vehicle to fly. I'm thinking this problem might have evolved organically. Perhaps getting all parameters up front was a much smaller ask a few years ago, but as more and more parameters are added, we just tack them into the package of what gets downloaded. 

I notice this most because I typically use MP on the bench and QGC when I'm flying. When using QGC I operate with far fewer parameters that I care about while at the field, whereas on the bench I definitely need  a lot more of them, often because MP needs them, but I"m pretty much always connected by USB when I use Mission Planner.

Would it be possible to narrow this down to an initial download of a much smaller package of "essential" parameters required for flying while connected over ELRS?

**Comment by wvarty** [comment]
I've pushed up https://github.com/ExpressLRS/ExpressLRS/pull/2626 on the ELRS side, for when this merges.

File: libraries/GCS_MAVLink/GCS_FTP.cpp [hunk 5]
```diff
@@ -155,9 +146,25 @@ void GCS_MAVLINK::ftp_error(struct pending_ftp &response, FTP_ERROR error) {
...
 {
+    ftp.last_send_ms = AP_HAL::millis(); // Used to detect active FTP session
+
     while (!send_ftp_reply(reply)) {
         hal.scheduler->delay(2);
     }
+
+    if (reply.req_opcode == FTP_OP::TerminateSession) {
+        ftp.last_send_ms = 0;
+    }
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_FTP.cpp:157]: Why move this?

File: libraries/GCS_MAVLink/GCS_FTP.cpp [hunk 6]
```diff
@@ -155,9 +146,25 @@ void GCS_MAVLINK::ftp_error(struct pending_ftp &response, FTP_ERROR error) {
 // send our response back out to the system
 void GCS_MAVLINK::ftp_push_replies(pending_ftp &reply)
 {
+    ftp.last_send_ms = AP_HAL::millis(); // Used to detect active FTP session
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_FTP.cpp:149]: Why move this?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 7]
```diff
@@ -827,6 +827,15 @@ float GCS_MAVLINK::telemetry_radio_rssi()
     return last_radio_status.rssi/254.0f;
 }
 
+bool GCS_MAVLINK::last_txbuf_is_greater(uint8_t txbuf_limit)
+{
+    if (AP_HAL::millis() - last_radio_status.received_ms > 5000) {
+        // stale report
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_Common.cpp:833]:
```suggestion
        // stale report, assume 100% buffer space
```

File: libraries/GCS_MAVLink/GCS.h [hunk 8]
```diff
@@ -759,6 +759,7 @@ class GCS_MAVLINK
         uint32_t remrssi_ms;
         uint8_t rssi;
         uint32_t received_ms; // time RADIO_STATUS received
+        uint8_t txbuf = 100;
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS.h:762]: With the new timeout we don't need to init to 100 anymore. 
```suggestion
        uint8_t txbuf;
```

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 9]
```diff
@@ -827,6 +827,15 @@ float GCS_MAVLINK::telemetry_radio_rssi()
     return last_radio_status.rssi/254.0f;
 }
 
+bool GCS_MAVLINK::last_txbuf_is_greater(uint8_t txbuf_limit)
+{
+    if (AP_HAL::millis() - last_radio_status.received_ms > 5000) {
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_Common.cpp:832]:
```suggestion
    if ((last_radio_status.received_ms == 0) || ((AP_HAL::millis() - last_radio_status.received_ms) > 5000)) {
```
Check for zero to make sure we do the correct thing if never received.

File: libraries/GCS_MAVLink/GCS.h (hunk 8 above)
Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS.h:762]: If we add the check for 0 ms.

**Review by IamPete1 — COMMENTED** [review]
Needs a rebase, its conflicting with something that has been merged recently.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Also needs a squash to remove the merge commit.

File: libraries/GCS_MAVLink/GCS_Param.cpp [hunk 10]
```diff
@@ -75,7 +75,7 @@ GCS_MAVLINK::queued_param_send()
     }
     count -= async_replies_sent_count;
 
-    while (count && _queued_parameter != nullptr && get_last_txbuf() > 50) {
+    while (count && _queued_parameter != nullptr && last_txbuf_is_greater(33)) {
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp:78]: Maybe a `static constexpr` for the magic number?

**Comment by tridge** [comment]
@brad112358 thanks! needs a rebase and squash. If you don't know how to do a git squash then ask on the code_review channel and I can do it

**Comment by IamPete1** [comment]
A tlog of a test on real hardware would be useful. With my SIM (https://github.com/ArduPilot/ardupilot/pull/26663) I'm seeing the FTP replies carry on for a long time after mission planner says the params have been fetched (like 30 seconds). It could be a mission planner issue of course, or we might need some fixes on top of this. I would guess mission planner gets impatient and requests reads multiple times resulting lots of duplicate requests.

**Comment by timtuxworth** [comment]
> It could be a mission planner issue of course, or we might need some fixes on top of this. I would guess mission planner gets impatient and requests reads multiple times resulting lots of duplicate requests.

It might be useful to test with QGC Peter. I've noticed that it behaves differently (sometimes better) and having that different behavior might give different/additional data to help figure out what is going on.

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 6 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_FTP.cpp:149]: Because we want to detect the active FTP session as soon as possible in get_reschedule_interval_ms(), even if we are blocked from sending the first response because of txbuf flow control.  This slows down normal streams sooner and allows the radio's buffer to drain more quickly and reduces the delay the GCS sees, which helps avoid GCS timout on the first FTP response.

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 5 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_FTP.cpp:157]: Because this does not need to be part of send_ftp_reply() and it is more clear when placed near other uses of ftplast_send_ms.

File: libraries/GCS_MAVLink/GCS.h (hunk 8 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS.h:762]: It seems more efficient to initialize just once than to execute the extra check every time.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 9 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp:832]: Would a comment pointing out the the initialization of txbuf accomplishes this be OK instead?

File: libraries/GCS_MAVLink/GCS_Param.cpp (hunk 10 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Param.cpp:78]: What would we call it?  one_third?  Is that really better?  Anything less obvious and I would be annoyed when reading the code to have to look up the constant.   I considered eliminating the parameter and hard code 33 in last_txbuf_is_greater(), but I wanted to allow for other threshold values, even though I doubt we will need them.

But I'll make the change if you let me know what I should call it.

**Comment by brad112358** [comment]
> @brad112358 thanks! needs a rebase and squash. If you don't know how to do a git squash then ask on the code_review channel and I can do it

Thanks.  I will learn how to do that and push again after recent review comments are resolved.

**Comment by brad112358** [comment]
@IamPete1 and @peterbarker If you could respond to my comments on your suggested changes, then I can clean this up.
Thanks!

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 6 above)
Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS_FTP.cpp:149]: OK, that makes sense. But its there now a risk of of FTP timing out if we sit here waiting to send for a while. Looks like the threshold is 3 seconds? Maybe we need a better way to check if there is FTP stuff happening than just looking at `last_send_ms`.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 11]
```diff
@@ -1142,7 +1151,7 @@ uint16_t GCS_MAVLINK::get_reschedule_interval_ms(const deferred_message_bucket_t
         // we are sending requests for waypoints, penalize streams:
         interval_ms *= 4;
     }
-    if (AP_HAL::millis() - ftp.last_send_ms < 500) {
+    if (AP_HAL::millis() - ftp.last_send_ms < 1000) {
```
Comment (IamPete1) [inline libraries/GCS_MAVLink/GCS_Common.cpp:1154]: It seems odd that we would use 1 second here but 3 seconds to time out a FTP session.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 9 above)
Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS_Common.cpp:832]: Yeah, comment is OK too. Really the key thing is to be able to understand what is going on without having to go look in the `.h`.

**Comment by IamPete1** [comment]
> @IamPete1 and @peterbarker If you could respond to my comments on your suggested changes, then I can clean this up. Thanks!

Any chance of that tlog? As I say even with this I didn't get good results (although still better than without), have you been testing with Mission planner or another GCS?

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 11 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp:1154]: I've only looked at QGroundControl which has a 1 second ack/nak timeout with 3 retries.  Does Mission Planner use a 3 second timeout?

Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS_Common.cpp:1154]: I was thinking of this timeout:
https://github.com/ArduPilot/ardupilot/blob/01b0e0c27cf91169fb00c5e18197e613539a5339/libraries/GCS_MAVLink/GCS_FTP.cpp#L35

**Comment by brad112358** [comment]
> Any chance of that tlog? As I say even with this I didn't get good results (although still better than without), have you been testing with Mission planner or another GCS?

I don't use ELRS so I've only tested with mLRS and briefly with SiK and Crossfire.
I've been using this change in the Copter-4.4 branch with QGroundControl and mLRS when the weather permits flying for a couple of months.  I've also tested with the master branch of Ardupilot on Mission Planner, QGC, and if I remember correctly, briefly with MAVProxy.

I'm traveling for the next few days, but I can capture tlogs with mLRS when I get back if I don't get to it sooner.

Perhaps I can also test with ELRS as I do have hardware it supports, but I expect that ELRS might require some work to get it working perfectly with FTP, even with this PR.

File: libraries/GCS_MAVLink/GCS_FTP.cpp (hunk 6 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_FTP.cpp:149]: Ultimately, since messages can be lost and the GCS may give up on FTP, testing for FTP activity will need to involve a timeout and I think this may be the simplest way.  Where is the timeout 3 seconds?  QGC seems to use 1 second timeouts.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 11 above)
Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp:1154]: (our responses crossed here)  OK.  I think things should work OK with this changed to 3 seconds.  The stream rates will just take 3 seconds to go back to normal instead of 1.  Should we move the define to a header file and use FTP_SESSION_TIMEOUT instead of 500 or 1000 in GCS_Common.cpp?

Comment (brad112358) [reply libraries/GCS_MAVLink/GCS_Common.cpp:1154]: Actually, after more consideration, I think there is little reason that this timeout should be the same as the session timeout.  This is essentially a heuristic and should be based on the expected behavior of the GCS, including radio latency, but doesn't need to match the session timeout which really needs to be conservative (long) as discarding the open state too early would be bad.   With the current GCS behavior, I think 1 second is fine.  
On the other hand, it would be easier for slower radios to make this work well if the GCS FTP retry timeout was a bit longer, say 2 seconds.  Bumping this from 500 to 1000 was enough for mLRS which was already working hard to keep buffered data to a minimum.  But, ELRS is even slower than mLRS.  Should we increase this to 2 seconds in anticipation that some work may also be required on the GCS side to help out ELRS?

**Comment by IamPete1** [comment]
> I don't use ELRS so I've only tested with mLRS and briefly with SiK and Crossfire.

Ah, OK, if you can give me a rough data rate for mLRS I can set that in the simulator and see how it gets on. I will also try out QGC.

Sorry for being slow, i'm keen to get https://github.com/ArduPilot/ardupilot/pull/26532 so we can see the tx and rx data rates which should make testing easyer.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 11 above)
Comment (IamPete1) [reply libraries/GCS_MAVLink/GCS_Common.cpp:1154]: My thinking is that if FTP thinks the session is still active then we should still lower its data rates. However a quick test shows the MAVProxy terminates the session after param download whereas QGC and MP do not. If we get the GCS to terminate correctly there should be no problem in using the longer time out here.

**Comment by brad112358** [comment]
The nominal down link data rates for mLRS are 1558, 2542, 4100, and 9102 bytes/sec.  Up link is slightly slower to allow separate RC data.  Data is sent in 82 byte chunks at 19, 31, 50, or 111 Hz.  There is also a mode which can compress data, but it doesn't make a big improvement and I haven't been using it.  My tests are all performed at the 19 Hz, 1558 bytes/sec rate, because I only have Frsky R9 hardware for mLRS and it is limited to this rate.  Other mLRS users have tested this PR for me with the other rates.  At the 1558 B/s rate, this PR improved parameter download time from about 50 seconds to about 6 seconds because FTP was able to be used with no lost messages and no retrys.  The higher rates saw less dramatic improvements.

Note, however, that  in mLRS, Radio Status messages are not always sent at a fixed rate and the values for txbuf do not represent actual percentage, but are fudged for Ardupilot or PX4 with the goal of using between 75 and 85 percent of the available data bandwidth.  This results in very low average buffer use in steady state conditions even though we allocate a fairly large, 2K byte buffer to allow clumps of messages.  Usually the buffer is empty by the time the next group of messages arrives.  This is quite different than a simple direct calculation of txbuf which seeks to keep the buffer utilization at around 50% full and results in higher latency than what is accomplished in mLRS, especially for very slow links.

My understanding was that ELRS at one point was using the same strategy as mLRS, but I see that your emulation is simply sending txbuf based on buffer used at a higher fixed rate so I'm not sure if ELRS discarded the ideas from mLRS at some point.  It is possible that the trouble you are seeing in your ELRS emulation is due to high latency because too much data is being buffered.  If so, increasing the timeouts in the GCS can help, but ELRS may also need to make changes to reduce buffered data related latency.

You can find the method mLRS uses [here](https://github.com/olliw42/mLRS/blob/main/mLRS/CommonRx/mavlink_interface_rx.h).

**Comment by IamPete1** [comment]
Rebased, removed the merge commit and squashed the comment type fixup commit.

**Comment by IamPete1** [comment]
Some testing with my sim set to a limit of 1500B/s.

Without this patch:
image
TxBuf is a 0 for the duration of the FTP. Average data rate is hard up against 1500. FTP takes 16 seconds.

With patch:
image

TxBus no longer hard on zero, average data rate is the same 1500. FTP takes 12 second.


In my testing FTP succeeded both times. When I tested before with a 500B/s data rate limit it failed both times. But there is probably somewhere in between where this could make the difference between FTP working or not and result in a much more significant improvement. 

I think this is a step in the right direction, but I suspect we will need to do more to get the most from the more limited links.

**Comment by brad112358** [comment]
To restate my goal, this PR is just providing radios with a way to make flow control work for FTP.  I think anything more is impractical given the limits of the information contained in the existing RADIO_STATUS message.

I never expected that this change would accomplish optimal FTP performance with most existing radio strategies for sending txbuf updates.  It just provides an opportunity (I used the phrase fighting chance for a reason) for even slow radios to make it work efficiently with some effort on the part of the radio developers.  There are a few possible approaches I have experimented with on the radio side which have been successful.  In the best case where the txbuf strategy on the radio side has already been designed to work well with existing Ardupilot code for non-FTP based bulk parameter download, FTP probably also works much better with this PR because it makes FTP respect txbuf the same way bulk parameter download currently does.  But, mLRS is the only radio I know of that has been completely successful in developing an effective strategy on the radio side.  Crossfire had some test releases that worked well for traditional bulk parameter download a couple of years ago, but they seemed to have dropped the effort when I stopped pestering them about it.  Radios like SiK which send radio status only at a fixed 1 Hz rate have no hope of efficiently supporting lumpy traffic like FTP even with this PR.  Until more radios can be improved to take advantage of the opportunity this PR is attempting to provide, we are stuck trying to match the serial baud rate with the radio link speed and hope the 1/3 utilization limit is effective.

But mLRS does work really well with this change and I think ELRS MAVLink mode can also easily be made to work quite well with this PR since it is still in development.  For the rest of the existing radios, my short term goal is only not to make things worse.  In the long term, they can be made more efficient as well if anyone cares enough to try.

What is the simulated link baud rate of your test?  When the baud rate is low enough compared to the radio throughput, FTP will probably work, though, not optimally because of the artificial 1/3 link speed rate limiting which Ardupilot now does even without txbuf flow control.  But, I normally use 230K baud for MAVLink with mLRS to minimize transmission based message latency (mLRS accumulates and parses every message at both ends of the link before starting to send it over the link, so a slow baud rate affects latency even more than it might otherwise).  Last I checked, ELRS was using a similar high baud rate, probably for the same reasons.  At these higher baud rates, FTP will fail completely without some form of flow control.  Also, what is the radio buffer size in your tests?  I assume you are not emulating what mLRS does, but just reporting raw txbuf percentages.  How often are you sending radio status? At higher baud rates you need to have a huge buffer (bad) or send status very frequently if sending at a fixed rate or (better) send an extra radio status report whenever the buffer nears full or empty to stop or start the flow immediately.  If the radio uses a good strategy for when to send radio status, you should be able achieve near full radio link utilization during FTP transfers without message loss with a reasonably small buffer.  But you need smarts in the radio in addition to something like this PR on the FC side.

**Comment by brad112358** [comment]
I have synced my local repository with the latest changes here (and from master), configured and built for one of my 19 Hz (1558 Bps) mLRS builds.  I have re-run a few quick sanity tests and uploaded tlogs covering the initial connection and parameter download for QGC, MP, and MAVProxy [here](https://drive.google.com/drive/folders/12HSouLG1FpzbuC206e4Y7HSbT6qCNZMl?usp=drive_link).  I have not analyzed these logs, but the parameter upload appeared to succeed quickly as expected in all 3 tests.

Sorry it took so long for me to get to this.

Please let me know what you gleam from these logs.

Thanks!

**Comment by brad112358** [comment]
In all these years, I've never bothered with SITL, but I guess I should figure out how to set it up so I can see your radio simulator work...

**Comment by brad112358** [comment]
Sorry, I see I forgot to enable general access to that google drive folder.  I've fixed that now.

**Comment by IamPete1** [comment]
The QGC log looks more or less as I would have expected. 

Looks like the mission planner on FTP failed and it grabbed params the old way. 

The MAVProxy one looks OK, it does FTP, but then we have a mission planner instance appear and grab param the old way.

**Comment by brad112358** [comment]
What are you using to examine the tlog files?

> Looks like the mission planner on FTP failed and it grabbed params the old way.

I'll try it again.  I didn't time it, bu it seemed to finish faster than I thought it would with bulk parameter transfer the old way (normally takes 5-8 times as long as FTP takes)

> The MAVProxy one looks OK, it does FTP, but then we have a mission planner instance appear and grab param the old way.

That is a bit of a mystery.  How do you know it was mission planner?  I used a serial to WiFi UDP bridge in AP mode connected to the Tx module for this test and MAVProxy was running on a Linux system which was supposed to be the only client connected to the bridge WiFi network and there is no mission planner installed there.  I did have my other notebook still booted into windows upstairs, but it was only connected to my home WiFi and not the bridge AP because it was downloading windows updates.  I do sometimes run apmplanner2 on that Linux box, but I didn't think it was still running at the time either.

SITL was surprisingly trivial to get running.  I created a local branch with both your radio simulator and this code but I'm not sure how to configure and enable the radio simulation and connect a second GS to it or how to make the first MAVProxy instance use the simulated radio link.  Can you give me some hints?

**Comment by IamPete1** [comment]
> What are you using to examine the tlog files?
> 
https://firmware.ardupilot.org/Tools/WebTools/StreamStats/

Its still very WIP, you'll probably want to turn down the window size for looking at param stuff.

The MAVProxy one looks like this:
image

We see the FTP in purple, takes a little over 10 seconds. We don't see it on the message rate plot (I think because there is only one point so there is no line) but the total plot shows a huge spike at the end of the FTP. Its too fast to actually have come over the data link I think, The top of the page shows that there were some param requests from mission planner, but they were coming from system ID 1 which is all a bit odd, maybe mission planner was connecting via UDP to the forwarding from MAVProxy or something like that:

image

We do see radio messages from sys ID 51.
image

So I guess that is the ground side radio. ELRs don't send on the ground side, so that must be something that MLRS does, although looks like we don't get buffer info, just noise stats.

Then finaly we see the GCS on 255, 
image

More or less what we would expect, some heartbeats, stream rate requests, commands and the FTP requests.

This tool does not show any of the actual data in the message, just what there is and the rates, but UAV plotter can take tlogs if you want to look inside the messages. https://plot.ardupilot.org/#/

**Comment by IamPete1** [comment]
> SITL was surprisingly trivial to get running. I created a local branch with both your radio simulator and this code but I'm not sure how to configure and enable the radio simulation and connect a second GS to it or how to make the first MAVProxy instance use the simulated radio link. Can you give me some hints?

You can add the "radio" to any serial port with the extra command `-A --serial2=sim:ELRS:tcp:3` In that example it adds it to serial 2 and then you connect on the standard serial 2 port which is 5763. 

SITL needs a GCS to run, so you would typically connect a second GCS instance to the second port.

**Comment by IamPete1** [comment]
We were having no luck getting past CI, so I have rebased to see if that will clear it.

**Comment by brad112358** [comment]
> Looks like the mission planner on FTP failed and it grabbed params the old way.
> The MAVProxy one looks OK, it does FTP, but then we have a mission planner instance appear and grab param the old way.

I ran both the MP and MAVProxy tests again and snooped on the messages with wireshark.  I didn't see any sign of what you observed above so I looked at the original tlogs again with StreamStats.  In both cases the parameter messages you noted are reported in a huge burst all at the exact time that the FTP transfer completed.  I haven't checked the source code to be sure, but I'm guessing these are just parameter details which MP and MAVProxy both place in the tlog and StreamStats is interpreting this data as messages.  I'm guessing QGC doesn't inject the parameter values in the tlog, or perhaps it uses a different format that doesn't show up in StreamStats.

So, I think everything looks as expected with all 3 GCS.

**Review by IamPete1 — APPROVED** [review]
LGTM
